tls/s2n_resume.c

```diff
@@ -769,187 +769,147 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
     struct s2n_stuffer aad = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&aad, &aad_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
 
-    POSIX_GUARD(s2n_stuffer_read_bytes(from, key_name, s2n_array_len(key_name)));
-
-    key = s2n_find_ticket_key(conn->config, key_name);
-
-    /* Key has expired; do full handshake with New Session Ticket (NST) */
-    S2N_ERROR_IF(!key, S2N_ERR_KEY_USED_IN_SESSION_TICKET_NOT_FOUND);
-
-    POSIX_GUARD(s2n_stuffer_read(from, &iv));
-
-    POSIX_GUARD(s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN));
-    POSIX_GUARD(s2n_session_key_alloc(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.init(&aes_ticket_key));
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.set_decryption_key(&aes_ticket_key, &aes_key_blob));
-
-    POSIX_GUARD(s2n_stuffer_init(&aad, &aad_blob));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
-
+    /* Initialize blob to be decrypted */
     struct s2n_blob en_blob = { 0 };
     uint32_t en_blob_size = s2n_stuffer_data_available(from);
     uint8_t *en_blob_data = s2n_stuffer_raw_read(from, en_blob_size);
-    POSIX_ENSURE_REF(en_blob_data);
-    POSIX_GUARD(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
-    POSIX_GUARD(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+    RESULT_ENSURE_REF(en_blob_data);
+    RESULT_GUARD_POSIX(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
 
+    RESULT_GUARD_POSIX(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+    RESULT_GUARD(s2n_aes256_gcm.destroy_key(&aes_ticket_key));
+
+    /* Parse decrypted state */
     struct s2n_blob state_blob = { 0 };
     uint32_t state_blob_size = en_blob_size - S2N_TLS_GCM_TAG_LEN;
-    POSIX_GUARD(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
+    RESULT_GUARD_POSIX(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
     struct s2n_stuffer state_stuffer = { 0 };
-    POSIX_GUARD(s2n_stuffer_init(&state_stuffer, &state_blob));
-    POSIX_GUARD(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
-    POSIX_GUARD_RESULT(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&state_stuffer, &state_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
+    RESULT_GUARD(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
 
-    if (s2n_connection_get_protocol_version(conn) >= S2N_TLS13) {
-        return S2N_SUCCESS;
+    if (s2n_connection_get_protocol_version(conn) >= S2N_TLS13 || conn->config->use_session_cache) {
+        return S2N_RESULT_OK;
     }
```

Comment:
Is the new "conn->config->use_session_cache" because s2n_resume_decrypt_session_cache is now using this method, and didn't include the rest of the logic in its copy/paste implementation?

Reply:
Yes, session caching doesn't need the NST handshake method so that last paragraph shouldn't be executed.